File: legacylift/server/core/migration/target_api.py
"""
core/migration/target_api.py — Builds the "ALREADY-MIGRATED TARGET API" block:
a compact, deterministic contract of the target code that has ALREADY been
generated for the units this chunk depends on (and its same-file siblings), so
the model calls the real generated names/signatures instead of inventing new
ones. This is the generated-side half of cross-chunk context (the source-side
half is the DIRECT DEPENDENCIES block built from related chunks' legacy source).

Server twin of client/lib/targetApi.ts. Never raises.
"""

from __future__ import annotations

from pathlib import PurePosixPath
from typing import Callable, Optional

from core.target_languages import get_target_language
from utils.symbol_index import extract_exports

# Keep the block well under the migrate request's manifest-style budget.
_MAX_CHARS = 9_000
_MAX_DEP_SOURCE_CHARS = 12_000


def render_dependency_source(
    dep_units: list[dict],
    *,
    max_chars: int = _MAX_DEP_SOURCE_CHARS,
) -> str:
    """Render the legacy SOURCE of the units this chunk calls (+ their rule) for
    the DIRECT DEPENDENCIES block. `dep_units` items: {name, source, rule}."""
    blocks: list[str] = []
    for unit in dep_units or []:
        source = (unit.get("source") or "").strip()
        if not source:
            continue
        parts = [f"--- {unit.get('name', '')} ---"]
        rule = (unit.get("rule") or "").strip()
        if rule:
            parts.append(f"rule: {rule}")
        parts.append(source)
        candidate = "\n".join(parts)
        if blocks and sum(len(b) + 2 for b in blocks) + len(candidate) > max_chars:
            break
        blocks.append(candidate)
    return "\n\n".join(blocks)


def _module_name(filename: str, language: str) -> str:
    """`interest_calc.cbl` + Python → `interest_calc.py`."""
    stem = PurePosixPath(filename or "").stem or "module"
    ext = get_target_language(language).extension
    return f"{stem}{ext}"


def build_target_api(
    project,
    chunk_id: str,
    resolve_language: Callable[[str], str],
    *,
    max_chars: int = _MAX_CHARS,
) -> str:
    """Render the already-generated target API relevant to `chunk_id`.

    Args:
        project: the live Project (reads layer0_chunks, layer0_graph,
            chunk_migrations, current_migration, chunk_approvals).
        chunk_id: the chunk about to be migrated.
        resolve_language: filename → target language name (e.g. per-file target).
    """
    layer0_chunks = getattr(project, "layer0_chunks", None) or []
    chunk_by_id: dict[str, dict] = {c.get("id"): c for c in layer0_chunks if c.get("id")}
    current = chunk_by_id.get(chunk_id)
    if current is None:
        return ""

    current_file = current.get("filename", "")

    # Generated code per chunk: approved store first, then the latest (draft) run.
    migrations: dict[str, str] = dict(getattr(project, "chunk_migrations", None) or {})
    current_migration = getattr(project, "current_migration", None) or {}
    if isinstance(current_migration, dict):
        cm_id = current_migration.get("chunk_id")
        cm_code = current_migration.get("migrated_code") or ""
        if cm_id and cm_id != chunk_id and cm_code.strip() and cm_id not in migrations:
            migrations[cm_id] = cm_code  # a generated-but-unapproved neighbour

    approvals: dict[str, str] = dict(getattr(project, "chunk_approvals", None) or {})

    # ── relevance ranking ──────────────────────────────────────────────────
    dependency_ids = _direct_dependency_ids(project, chunk_id, set(chunk_by_id))
    sibling_ids = [
        cid for cid, c in chunk_by_id.items()
        if cid != chunk_id and c.get("filename", "") == current_file
    ]

    ordered_ids: list[str] = []
    seen: set[str] = {chunk_id}
    for group in (dependency_ids, sibling_ids, list(chunk_by_id)):
        for cid in group:
            if cid not in seen and (migrations.get(cid) or "").strip():
                ordered_ids.append(cid)
                seen.add(cid)

    # ── render ─────────────────────────────────────────────────────────────
    lines: list[str] = []
    for cid in ordered_ids:
        chunk = chunk_by_id[cid]
        code = migrations.get(cid, "")
        filename = chunk.get("filename", "")
        language = resolve_language(filename) if filename else resolve_language(current_file)
        surface = extract_exports(code, language)
        if surface.is_empty():
            continue

        status = "approved" if approvals.get(cid) == "approved" else "draft"
        header = f"- {chunk.get('name', cid)}  →  {_module_name(filename, language)}  [{status}]"
        block = [header, *surface.as_lines()]
        candidate = "\n".join(block)

        if lines and sum(len(x) + 1 for x in lines) + len(candidate) > max_chars:
            break
        lines.append(candidate)

    return "\n".join(lines)


def _direct_dependency_ids(project, chunk_id: str, id_set: set[str]) -> list[str]:
    """Chunk ids this chunk directly calls (resolved graph `call` edges)."""
    graph = getattr(project, "layer0_graph", None) or {}
    out: list[str] = []
    for edge in graph.get("edges", []) or []:
        if not isinstance(edge, dict):
            continue
        if edge.get("edge_type") in ("data_read", "data_write"):
            continue
        if edge.get("source") == chunk_id:
            tgt = edge.get("target")
            if tgt in id_set and tgt != chunk_id and tgt not in out:
                out.append(tgt)
    return out

File: legacylift/client/lib/targetApi.ts
// lib/targetApi.ts - Builds the "ALREADY-MIGRATED TARGET API" block: a compact,
// deterministic contract of the target code that has ALREADY been generated for
// the units a chunk depends on (and its same-file siblings), so the model calls
// the real generated names/signatures instead of inventing new ones. This is the
// generated-side half of cross-chunk context; the source-side half is the DIRECT
// DEPENDENCIES block (buildDependenciesSource), the legacy source of the units
// this chunk calls.
//
// Client twin of server/core/migration/target_api.py.

import type {
  BusinessRule,
  DependencyGraph,
  MigrationChunk,
} from "@/types/legacylift";
import type { TargetLanguage } from "@/lib/targetLanguages";
import { extractExports, isEmptySurface, surfaceLines } from "@/lib/symbols";

const MAX_API_CHARS = 9_000;
const MAX_DEP_SOURCE_CHARS = 12_000;

/** filename without its extension: "interest_calc.cbl" -> "interest_calc". */
function stem(filename: string): string {
  const base = filename.split(/[\\/]/).pop() ?? filename;
  const dot = base.lastIndexOf(".");
  return dot > 0 ? base.slice(0, dot) : base;
}

/** Resolve a graph key (node id OR unit name) to a chunk - tolerates both the
 *  offline (name-keyed) and backend (id-keyed) client graph shapes. */
function makeResolver(chunks: MigrationChunk[]): (key: unknown) => MigrationChunk | undefined {
  const byId = new Map(chunks.map((c) => [c.id, c] as const));
  const byName = new Map(chunks.map((c) => [c.name, c] as const));
  return (key) =>
    typeof key === "string" ? byId.get(key) ?? byName.get(key) : undefined;
}

function isDataEdge(edge: { label?: string; edge_type?: string }): boolean {
  const kind = (edge.label ?? edge.edge_type ?? "").toLowerCase();
  return kind.includes("data");
}

/** The chunks this chunk directly calls (resolved graph call edges). */
export function directDependencies(
  chunks: MigrationChunk[],
  graph: DependencyGraph | null,
  current: MigrationChunk,
): MigrationChunk[] {
  const resolve = makeResolver(chunks);
  const out: MigrationChunk[] = [];
  const seen = new Set<string>();
  for (const edge of graph?.edges ?? []) {
    if (isDataEdge(edge as { label?: string; edge_type?: string })) continue;
    const src = resolve(edge.source);
    if (!src || src.id !== current.id) continue;
    const tgt = resolve(edge.target);
    if (tgt && tgt.id !== current.id && !seen.has(tgt.id)) {
      seen.add(tgt.id);
      out.push(tgt);
    }
  }
  return out;
}

/** Render an ordered list of already-generated units as compact API blocks. */
function renderApiBlocks(
  units: MigrationChunk[],
  resolveTarget: (filename: string) => TargetLanguage,
  maxChars: number,
): string {
  const blocks: string[] = [];
  for (const c of units) {
    if (!(c.migrated_code ?? "").trim()) continue;
    const target = resolveTarget(c.source_file);
    const surface = extractExports(c.migrated_code, target.language);
    if (isEmptySurface(surface)) continue;

    const status = c.status === "Approved" ? "approved" : "draft";
    const header = `- ${c.name}  ->  ${stem(c.source_file)}${target.extension}  [${status}]`;
    const candidate = [header, ...surfaceLines(surface)].join("\n");

    if (blocks.length && blocks.join("\n").length + candidate.length > maxChars) break;
    blocks.push(candidate);
  }
  return blocks.join("\n");
}

/**
 * Render the already-generated target API relevant to `current`: its direct
 * dependencies first, then same-file siblings, then any other generated unit,
 * budget-capped. Only units that already have generated code are included.
 */
export function buildTargetApi(
  chunks: MigrationChunk[],
  graph: DependencyGraph | null,
  current: MigrationChunk,
  resolveTarget: (filename: string) => TargetLanguage,
  maxChars = MAX_API_CHARS,
): string {
  const deps = directDependencies(chunks, graph, current);
  const siblings = chunks.filter(
    (c) => c.id !== current.id && c.source_file === current.source_file,
  );

  const ordered: MigrationChunk[] = [];
  const seen = new Set<string>([current.id]);
  for (const group of [deps, siblings, chunks]) {
    for (const c of group) {
      if (!seen.has(c.id) && (c.migrated_code ?? "").trim()) {
        ordered.push(c);
        seen.add(c.id);
      }
    }
  }

  return renderApiBlocks(ordered, resolveTarget, maxChars);
}

/**
 * Render the already-generated target API of units in OTHER files, so a file
 * being finalized reconciles its cross-file references to real neighbour
 * names/signatures rather than only fixing within-file drift.
 */
export function buildCrossFileApi(
  chunks: MigrationChunk[],
  resolveTarget: (filename: string) => TargetLanguage,
  currentFilename: string,
  maxChars = MAX_API_CHARS,
): string {
  const others = chunks.filter(
    (c) => c.source_file !== currentFilename && (c.migrated_code ?? "").trim(),
  );
  return renderApiBlocks(others, resolveTarget, maxChars);
}

/**
 * Render the legacy SOURCE of the units this chunk calls (+ their extracted
 * rule) for the DIRECT DEPENDENCIES block - the source-side half of "both".
 */
export function buildDependenciesSource(
  chunks: MigrationChunk[],
  graph: DependencyGraph | null,
  current: MigrationChunk,
  rules: BusinessRule[],
  maxChars = MAX_DEP_SOURCE_CHARS,
): string {
  const ruleForChunk = (c: MigrationChunk): string => {
    const r =
      rules.find((x) => x.chunk_id === c.id) ??
      rules.find((x) => x.source_file === c.source_file && x.title === c.name);
    return r?.description ?? "";
  };

  const blocks: string[] = [];
  for (const dep of directDependencies(chunks, graph, current)) {
    const source = (dep.source_code ?? "").trim();
    if (!source) continue;
    const parts = [`--- ${dep.name} ---`];
    const rule = ruleForChunk(dep).trim();
    if (rule) parts.push(`rule: ${rule}`);
    parts.push(source);
    const candidate = parts.join("\n");
    if (blocks.length && blocks.join("\n\n").length + candidate.length > maxChars) break;
    blocks.push(candidate);
  }
  return blocks.join("\n\n");
}
